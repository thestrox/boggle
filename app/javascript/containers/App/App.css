.root-container {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 50vh;
    width: 50vw;
}