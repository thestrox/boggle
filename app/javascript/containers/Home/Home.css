.home-container {
    display: flex;
    flex-direction: row;
}

.left-container {
    margin-right: 50px;
}

.board-container, .wordlist-container  {
    min-height: 200px;
}

.wordlist-container {
    height: 250px;
}

.word-input {
    margin-bottom: 10px;
}

.reset-button, .end-button, .submit-button, .new-button {
    margin-right: 10px !important;
}

.button-wrapper {
    margin-top: 10px !important;
}