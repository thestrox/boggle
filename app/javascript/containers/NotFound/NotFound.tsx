import React from "react";

export default function NotFound() {
    return (
        <div>
            <h1>404</h1>
            <h3>Page not found</h3>
        </div>
    )
}
