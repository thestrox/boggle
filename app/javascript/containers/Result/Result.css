.scorelist-container {
    min-width: 200px;
    min-height: 200px;
    max-height: 500px;
}

.total-container, .time-container {
    display: flex;
    padding: 8px;
    justify-content: space-between;
}