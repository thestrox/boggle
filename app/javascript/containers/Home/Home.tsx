import React from "react";
import Board from "../../components/Board/Board";
import { connect } from "react-redux";
import { AppState, AppThunk } from "../../store";
import { initializeNewBoard, validateWord, resetBoard } from "../../store/thunks";
import WordList from "../../components/WordList/WordList";
import { ValidateRequest } from "../../shared/types/api-types";
import { toast } from "react-toastify";
import Countdown, { zeroPad } from "react-countdown";
import { Redirect } from "react-router-dom";
import Button from '@material-ui/core/Button';
import "./Home.css"
import { Input } from "@material-ui/core";

export type HomeProps = {
    board: string[][];
    wordScoreMap: { [key: string]: number }
    duration: number;
    gameStartDate: number;
    initializeNewBoard: () => AppThunk,
    validateWord: (validateRequestBody: ValidateRequest) => AppThunk
    resetBoard: () => AppThunk
}

export type HomeState = {
    currentWord: string;
    toResult: boolean;
};

class Home extends React.Component<HomeProps, HomeState> {

    constructor(props: HomeProps) {
        super(props);
        this.state = {
            currentWord: '',
            toResult: false
        };
    }

    componentDidMount() {
        this.props.initializeNewBoard();
    }

    reset = () => {
        this.props.resetBoard();
    }

    newGame = () => {
        this.props.initializeNewBoard();
    }

    endGame = () => {
        this.setState({
            toResult: true
        });
    }

    validateWord = () => {
        if (this.props.wordScoreMap[this.state.currentWord]) {
            toast.error('Duplicate word. Please try again.');
        } else {
            this.props.validateWord({ board: this.props.board, word: this.state.currentWord });
        }
        this.updateCurrentWord("");
    }

    onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        this.updateCurrentWord(event.target.value);
    }

    updateCurrentWord = (word: string) => {
        this.setState({ currentWord: word.toUpperCase() });
    }

    keyPress = (e: React.KeyboardEvent<any>) => {
        if (e.key === "Enter") {
            this.validateWord();
        }
    }

    onTileClick = (row: number, col: number) => {
        this.updateCurrentWord(this.state.currentWord + this.props.board[row][col]);
    }

    render() {
        if (this.state.toResult) {
            return (
                <Redirect to='/result' />
            );
        }

        return (
            <div className="home-container">
                <div className="left-container">
                    <Input className="word-input"
                        value={this.state.currentWord}
                        onChange={this.onChange}
                        onKeyPress={this.keyPress}
                        placeholder="Type a suitable word"
                    />

                    <Board board={this.props.board} onTileClick={this.onTileClick} className="board-container"></Board>
                    <div className="button-wrapper">
                        <Button classes={{ root: 'submit-button' }} variant="contained" color="primary" onClick={this.validateWord}>Submit Word</Button>
                        <Button classes={{ root: 'end-button' }} variant="contained" color="default" onClick={this.endGame}>End Game</Button>
                    </div>

                </div>
                <div className="right-container">
                    {this.props.gameStartDate &&
                        <Countdown
                            date={this.props.gameStartDate + this.props.duration}
                            onComplete={this.endGame}
                            renderer={(props) => <span>{zeroPad(props.minutes)}:{zeroPad(props.seconds)}</span>}
                        />
                    }
                    <div>Results: </div>
                    <WordList wordScoreMap={this.props.wordScoreMap} className="wordlist-container"></WordList>
                    <div className="button-wrapper">
                        <Button classes={{ root: 'reset-button' }} variant="contained" onClick={this.reset}>Reset</Button>
                        <Button classes={{ root: 'new-button' }} variant="contained" onClick={this.newGame}>New Game</Button>
                    </div>


                </div>
            </div >
        );
    }
}

const mapStateToProps = (state: AppState) => {
    return {
        board: state.board.board,
        wordScoreMap: state.score.wordScoreMap,
        duration: state.board.duration,
        gameStartDate: state.board.gameStartDate,
    }
}

const HomeConnected = connect(
    mapStateToProps,
    {
        initializeNewBoard: initializeNewBoard,
        validateWord: validateWord,
        resetBoard: resetBoard
    }
)(Home)
export default HomeConnected