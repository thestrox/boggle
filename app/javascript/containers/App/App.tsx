import React from "react";
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'
import NotFound from "../NotFound/NotFound";
import Home from "../Home/Home";
import Result from "../Result/Result";
import "./App.css"

export default function App() {
    return (
        <div className='root-container'>
            <Switch>
                <Route exact path="/result" component={Result} />
                <Route exact path="/" component={Home} />
                <Route component={NotFound} />
            </Switch>
        </div>
    )
}