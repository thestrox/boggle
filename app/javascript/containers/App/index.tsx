import React from "react";
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'
import NotFound from "../NotFound";
import Home from "../Home";
import Result from "../Result";

export default function App() {
    return (
        <>
            <Switch>
                <Route exact path="/result" component={Result} />
                <Route exact path="/" component={Home} />
                <Route component={NotFound} />
            </Switch>
        </>
    )
}