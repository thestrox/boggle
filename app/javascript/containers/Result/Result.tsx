import React from "react";
import { connect } from "react-redux";
import { AppState } from "../../store";
import WordList from "../../components/WordList/WordList";
import "./Result.css"
import { Button } from "@material-ui/core";
import { formatTimeDelta, calcTimeDelta, zeroPad } from "react-countdown";
import { Redirect } from "react-router-dom";
import { reset } from "../../store/common-actions";

export type ResultProps = {
    wordScoreMap: { [key: string]: number };
    gameStartDate: number;
    resetBoard: () => void;
}

export type ResultState = {
    toHome: boolean;
}

class Result extends React.Component<ResultProps, ResultState> {
    constructor(props: ResultProps) {
        super(props);
        this.state = {
            toHome: this.props.gameStartDate ? false: true
        };

    }

    toHome = () => {
        this.props.resetBoard();
        this.setState({ toHome: true })
    }

    render() {
        if (this.state.toHome) {
            return (
                <Redirect to='' />
            );
        }

        const values = Object.values(this.props.wordScoreMap)
        const totalScore = values.length > 0 
        ? values.reduce((p, c) => p + c) 
        : 0;

        const formatted = formatTimeDelta(calcTimeDelta(Date.now() - this.props.gameStartDate, { controlled: true }))

        return (
            <div className="result-container">
                <span>Your Result</span>
                <WordList wordScoreMap={this.props.wordScoreMap} className="scorelist-container"></WordList>
                <div className="total-container">
                    <span>Total Score:</span>
                    <span>{totalScore}</span>
                </div>
                <div className="time-container">
                    <span>Time Elapsed:</span>
                    <span>{zeroPad(formatted.minutes)}:{zeroPad(formatted.seconds)}</span>
                </div>
                <Button onClick={this.toHome} variant="contained" color="primary">Home</Button>

            </div>
        );
    }
}

const mapStateToProps = (state: AppState) => {
    return {
        wordScoreMap: state.score.wordScoreMap,
        gameStartDate: state.board.gameStartDate
    }
}

const ResultConnected = connect(
    mapStateToProps,
    {resetBoard: reset}
)(Result)
export default ResultConnected
