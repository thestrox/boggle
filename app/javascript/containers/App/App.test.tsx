import React from 'react';
import { shallow } from 'enzyme';
import App from "./App"
import { Route } from 'react-router-dom';
import Result from "../Result/Result";
import NotFound from "../NotFound/NotFound";
import Home from "../Home/Home";

function setup() {
    return {
        component: shallow(<App/>)
    }
}

describe('App Component', () => {
    it('should render correctly', () => {
        const {component} = setup();
        expect(component.find('div').hasClass('root-container')).toBe(true);
        let routeNode = component.find(Route);
        expect(routeNode.at(0).prop('path')).toBe('/result');
        expect(routeNode.at(1).prop('path')).toBe('/');
    });

    it('should test snapshot renders', () => {
        const {component} = setup();
        expect(component).toMatchSnapshot();
    });
})