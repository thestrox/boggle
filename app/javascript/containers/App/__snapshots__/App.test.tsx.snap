// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should test snapshot renders 1`] = `
<div
  className="root-container"
>
  <Switch>
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Result)",
          "type": [Function],
        }
      }
      exact={true}
      path="/result"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Home)",
          "type": [Function],
        }
      }
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
    />
  </Switch>
</div>
`;
